### DISCUSSION TITLE
Analog dimmable LED driver solution - DAC

### DISCUSSION THREAD

**Discussion by bzgec** [stevenbennett/DIYson#48] [Ideas]
In video [The DIYson Electronics Pushed My Limits - DIYson Lamp Build Log #7](https://youtu.be/Emohi3FtTuM?list=PL0szyq6FLzZi2BB8-iA09LrTv0rd4K04u) you mention light flickering and possible solution using LED driver which uses analog value for dimming. You could use SAMD21 internal DAC to set the desired analog voltage value. I think that it should be possible, here is some link to get you started - https://learn.adafruit.com/circuitpython-essentials/circuitpython-analog-out.

**Comment by stevenbennett** [comment]
@bzgec Thanks for the suggestion and resources. I've actually tried getting this to work but was unsuccessful. I will give it another shot sometime. If anyone could do a proof-of-concept using the same parts that would be a huge help!

**Comment by bzgec** [comment]
I rewrote LED controller with Arduino framework - [repo](https://github.com/bzgec/DIYson-firmware) - [Arduino file](https://github.com/bzgec/DIYson-firmware/blob/master/DIYson.ino).
Note that I changed the pins you used because I have [Seeed Studio XIAO SAMD21](https://wiki.seeedstudio.com/Seeeduino-XIAO/) available and [QT Py SAMD21](https://www.adafruit.com/qtpy) is not fully pin compatible.

You will probably need to change the Touch configuration, because I only tried with jumper wires on breadboard (check `Touch filtering configuration` macros).
For this reason I created a `DEBUG_TOUCH_VALUES` macro. When defined, MCU prints raw and filtered touch values, which can be displayed with [Arduino's serial plotter](https://docs.arduino.cc/software/ide-v2/tutorials/ide-v2-serial-plotter) (for this option you need to use different pins - you used TX and RX pins for Touch pin and PWM).

**Reply by stevenbennett** [reply]
@bzgec This is great! Thank you so much for working on this and for putting so much thought and effort into the comments in the .ino file. I should be able to try this out using the QT Py - SAMD21 thanks to your documentation. Thank you!

How has your testing using the XIAO SAMD21 gone? Any issues? Any reason why I couldn't achieve your results using circuit python if I wanted to? (just out of curiosity).

**Reply by bzgec** [reply]
I didn't encounter any significant issues, aside from fine-tuning the touch configuration to ensure optimal performance. 
Since I'm accustomed to programming embedded systems in C, I didn't attempt to use CircuitPython. Exploring CircuitPython would require a considerable amount of time for me, and there might also be concerns regarding its execution speed.

**Reply by stevenbennett** [reply]
Totally understandable. It's probably best to use Arduino code anyway since it's so common. Looking forward to trying this out and possibly merging your solution with the main project. If this eliminates driver whine it will be a big step forward. Thank you again for your work, I really appreciate it. Any chance you are on the Discord server?

**Reply by stevenbennett** [reply]
@bzgec Which LED driver are you using by the way?

**Reply by bzgec** [reply]
I have just joined the Discord server! :)

Currently, I'm not using any LED driver or assembling the lamp. I thought I could lend a hand. Perhaps in the future, I'll tackle the DIYson project.
